lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
 
-    // If we have an object, we will not have live range issues, just continue.
-    if (arg->getType().isObject())
+  for (unsigned i : indices(argList)) {
+    Operand &argOp = argList[i];
+    SILParameterInfo param = paramList[i];
+    if (param.isIndirectMutating())
       continue;
 
-    // Now that we know that we have a non-argument address, perform a take-init
-    // of arg into a lifetime extended alloc_stack
-    args[i] = createLifetimeExtendedAllocStack(builder, loc, arg, exitingBlocks,
-                                               callbacks);
+    SILType ty = argOp.get()->getType();
+    if (!ty.isTrivial(*F) || (includeTrivialAddrArgs && ty.isAddress()))
+      argOperands.push_back(&argOp);
   }
+}
 
-  // Emit a destroy for each captured closure argument at each final release
-  // point.
-  for (auto *finalRelease : tracker.getFinalReleases()) {
-    builder.setInsertionPoint(finalRelease);
-    builder.setCurrentDebugScope(finalRelease->getDebugScope());
-    for (unsigned i : indices(args)) {
-      SILValue arg = args[i];
-      SILParameterInfo param = params[i];
-
-      releasePartialApplyCapturedArg(builder, loc, arg, param, callbacks);
+bool swift::collectDestroys(SingleValueInstruction *inst,
+                            SmallVectorImpl<SILInstruction *> &destroys) {
+  bool isDead = true;
+  for (Operand *use : inst->getUses()) {
+    SILInstruction *user = use->getUser();
+    if (useHasTransitiveOwnership(user)) {
+      if (!collectDestroys(cast<SingleValueInstruction>(user), destroys))
+        isDead = false;
+      destroys.push_back(user);
+    } else if (useDoesNotKeepValueAlive(user)) {
+      destroys.push_back(user);
+    } else {
+      isDead = false;
     }
   }
-
-  return true;
+  return isDead;
 }
 
-static bool
-deadMarkDependenceUser(SILInstruction *inst,
-                       SmallVectorImpl<SILInstruction *> &deleteInsts) {
-  if (!isa<MarkDependenceInst>(inst))
+/// Move the original arguments of the partial_apply into newly created
+/// temporaries to extend the lifetime of the arguments until the partial_apply
+/// is finally destroyed.
+///
+/// TODO: figure out why this is needed at all. Probably because of some
+///       weirdness of the old retain/release ARC model. Most likely this will
+///       not be needed anymore with OSSA.
+static bool keepArgsOfPartialApplyAlive(PartialApplyInst *pai,
+                                        ArrayRef<SILInstruction *> paiUsers,
+                                        SILBuilder &builder) {
+  SmallVector<Operand *, 8> argsToHandle;
+  getConsumedPartialApplyArgs(pai, argsToHandle,
+                              /*includeTrivialAddrArgs*/ false);
+  if (argsToHandle.empty())
+    return true;
+
+  // Compute the set of endpoints, which will be used to insert destroys of
+  // temporaries. This may fail if the frontier is located on a critical edge
+  // which we may not split.
+  ValueLifetimeAnalysis vla(pai, paiUsers);
```

Comment:
Thank you for fixing this. I was worried about this code using ValueLifetimeAnalysis(SILValue) API. We should really get rid of that entry point since it is so fragile and easy to mess up.